### PR TITLE
encrypting exploding chat messages

### PR DESCRIPTION
**Pull request by oconnor663**
cc @mmaxim and @joshblum. I don't have any tests in place for this yet, so there will probably be bugs, but it's at least all compiling if you're interested in looking at it now.

### CODE REVIEW

File: go/chat/teams.go [hunk 1]
```diff
@@ -237,6 +237,24 @@ func (t *TeamsNameInfoSource) DecryptionKeys(ctx context.Context, name string, t
 	return t.makeNameInfo(ctx, team, public)
 }
 
+func (t *TeamsNameInfoSource) EphemeralEncryptionKey(ctx context.Context, tlfName string, tlfID chat1.TLFID,
+	membersType chat1.ConversationMembersType, public bool) (keybase1.TeamEk, error) {
+	teamID, err := keybase1.TeamIDFromString(tlfID.String())
+	if err != nil {
+		return keybase1.TeamEk{}, err
```
Comment (joshblum) [inline go/chat/teams.go]: i'd just use named return args or make it a pointer if you want to explicits return `nil`

Comment (joshblum) [reply go/chat/teams.go]: why do we need so many args? should we be using the `membersType` since this won't work on `KBFS` type chats?

**Comment by oconnor663** [comment]
@joshblum and @mmaxim, this is ready for review.

File: go/chat/boxer.go [hunk 2]
```diff
@@ -1100,7 +1117,29 @@ func (b *Boxer) BoxMessage(ctx context.Context, msg chat1.MessagePlaintext,
...
+	// doesn't support exploding messages.
+	var ephemeralSeed *keybase1.TeamEk
+	if msg.IsExploding() {
+		ek, err := CtxKeyFinder(ctx, b.G()).EphemeralKeyForEncryption(
+			ctx, tlfName, msg.ClientHeader.Conv.Tlfid, membersType, msg.ClientHeader.TlfPublic)
+		if err != nil {
+			return nil, NewBoxingCryptKeysError(err)
+		}
+		ephemeralSeed = &ek
+		if version == chat1.MessageBoxedVersion_V1 {
```
Comment (joshblum) [inline go/chat/boxer.go]: can you move this up as a first check? we can just short circuit the keygen machinery if so

File: go/chat/boxer.go [hunk 3]
```diff
@@ -1256,23 +1296,39 @@ func (b *Boxer) boxV1(messagePlaintext chat1.MessagePlaintext, key types.CryptKe
...
 
+	// Regular messages use the same encryption key for the header and for the
+	// body. Exploding messages use a derived ephemeral key for the body.
+	bodyEncryptionKey := headerEncryptionKey
+	if messagePlaintext.IsExploding() {
+		bodyEncryptionKey, err = libkb.DeriveFromSecret(ephemeralSeed.Seed, libkb.DeriveReasonTeamEKExplodingChat)
+		if err != nil {
+			return nil, err
+		}
+		// The MessagePlaintext supplied by the caller has a Lifetime, but we
```
Comment (joshblum) [inline go/chat/boxer.go]: etime now not lifetime ;)

File: go/chat/boxer_test.go [hunk 4]
```diff
@@ -1488,3 +1499,71 @@ func TestRemarshalBoxed(t *testing.T) {
...
+	if typ, _ := body.MessageType(); typ != chat1.MessageType_TEXT {
+		t.Errorf("body type: %d, expected %d", typ, chat1.MessageType_TEXT)
+	}
+	if body.Text().Body != text {
+		t.Errorf("body text: %q, expected %q", body.Text().Body, text)
+	}
+	require.Nil(t, unboxed.SenderDeviceRevokedAt, "message should not be from revoked device")
+	require.NotNil(t, unboxed.BodyHash)
+	require.True(t, unboxed.IsExploding())
+	require.NotNil(t, unboxed.EphemeralMetadata())
```
Comment (joshblum) [inline go/chat/boxer_test.go]: i'd sanity check equality to the metadata you boxed here

**Comment by oconnor663** [comment]
`TestBackgroundPurge` (which I just rebased in) is failing, because it's now trying to use an ephemeral key where there is none. I'll add a fix.

File: go/chat/boxer.go [hunk 5]
```diff
@@ -52,7 +52,7 @@ type Boxer struct {
 	utils.DebugLabeler
 	globals.Contextified
 
-	boxWithVersion chat1.MessageBoxedVersion
```
Comment (joshblum) [inline go/chat/boxer.go]: inside of `compareHeadersMBV2` can you change the equality check on the metadata to only happen if we're on V3? just a sanity check if we're only supporting the `EphemeralMetadata` on v3. (sorry this was leftover from my purge PR but i didn't have the v3 stuff)

https://github.com/keybase/client/pull/11420/files#diff-d1325d9a259c7d25092218f3fa0661d4R868

File: go/chat/boxer_test.go [hunk 6]
```diff
@@ -1488,3 +1499,71 @@ func TestRemarshalBoxed(t *testing.T) {
...
+	// We need a user for unboxing to work.
+	u, err := kbtest.CreateAndSignupFakeUser("unbox", tc.G)
+	if err != nil {
+		t.Fatal(err)
+	}
+	msg := textMsgWithSender(t, text, gregor1.UID(u.User.GetUID().ToBytes()), chat1.MessageBoxedVersion_V3)
+
+	// Set the ephemeral metadata, to indicate that the messages is exploding.
+	msg.ClientHeader.EphemeralMetadata = &chat1.MsgEphemeralMetadata{
+		Lifetime: 99999,
```
Comment (joshblum) [inline go/chat/boxer_test.go]: i just landed https://github.com/keybase/client/pull/11439 so you should rebase to get the changes

File: go/chat/boxer.go (hunk 3 above)
Comment (oconnor663) [reply go/chat/boxer.go]: I'll leave it as is given our design talk this morning.

File: go/chat/boxer_test.go (hunk 4 above)
Comment (oconnor663) [reply go/chat/boxer_test.go]: Added an equality lifetime check. (The generation field is set by boxing and won't be equal.)

**Pull request closed without merging**